src/main/java/games/strategy/engine/lobby/server/login/LobbyLoginValidator.java

```diff
@@ -26,49 +40,64 @@
 import games.strategy.util.Version;
 
 public class LobbyLoginValidator implements ILoginValidator {
-  static final String THATS_NOT_A_NICE_NAME = "That's not a nice name.";
-  private static final String YOU_HAVE_BEEN_BANNED = "You have been banned from the TripleA lobby.";
-  private static final String USERNAME_HAS_BEEN_BANNED = "This username is banned, please create a new one.";
-  private static final String UNABLE_TO_OBTAIN_MAC = "Unable to obtain mac address.";
-  private static final String INVALID_MAC = "Invalid mac address.";
-  private static final Logger s_logger = Logger.getLogger(LobbyLoginValidator.class.getName());
   public static final String LOBBY_VERSION = "LOBBY_VERSION";
   public static final String REGISTER_NEW_USER_KEY = "REGISTER_USER";
   public static final String ANONYMOUS_LOGIN = "ANONYMOUS_LOGIN";
   public static final String LOBBY_WATCHER_LOGIN = "LOBBY_WATCHER_LOGIN";
   public static final String HASHED_PASSWORD_KEY = "HASHEDPWD";
+  public static final String ENCRYPTED_PASSWORD_KEY = "RSAPWD";
+  public static final String RSA_PUBLIC_KEY = "RSAPUBLICKEY";
   public static final String EMAIL_KEY = "EMAIL";
   public static final String SALT_KEY = "SALT";
-
-  public LobbyLoginValidator() {}
+  public static final String RSA = "RSA";
+  public static final String RSA_ECB_OAEPP = RSA + "/ECB/OAEPPadding";
+  static final String THATS_NOT_A_NICE_NAME = "That's not a nice name.";
+  private static final String YOU_HAVE_BEEN_BANNED = "You have been banned from the TripleA lobby.";
+  private static final String USERNAME_HAS_BEEN_BANNED = "This username is banned, please create a new one.";
+  private static final String UNABLE_TO_OBTAIN_MAC = "Unable to obtain mac address.";
+  private static final String INVALID_MAC = "Invalid mac address.";
+  private static final Logger logger = Logger.getLogger(LobbyLoginValidator.class.getName());
+  private static final Map<String, PrivateKey> rsaKeyMap = new HashMap<>();
 
   @Override
   public Map<String, String> getChallengeProperties(final String userName, final SocketAddress remoteAddress) {
     // we need to give the user the salt key for the username
     final Map<String, String> rVal = new HashMap<>();
-    final HashedPassword password = new DbUserController().getPassword(userName);
+    final HashedPassword password = new DbUserController().getLegacyPassword(userName);
     if (password != null && Strings.emptyToNull(password.value) != null) {
       rVal.put(SALT_KEY, MD5Crypt.getSalt(MD5Crypt.MAGIC, password.value));
     }
+    try {
+      final KeyPairGenerator keyGen = KeyPairGenerator.getInstance(RSA);
+      keyGen.initialize(4096);
+      final KeyPair keyPair = keyGen.generateKeyPair();
+      final String publicKey = Base64.getEncoder().encodeToString(keyPair.getPublic().getEncoded());
+      rVal.put(RSA_PUBLIC_KEY, publicKey);
+      rsaKeyMap.put(publicKey, keyPair.getPrivate());
```

Comment:
After a login attempt (successful or not), the key is being removed in [this line](https://github.com/RoiEXLab/triplea/blob/4f033dfa90695c8760f6ad1c47c777aa5b251b13/src/main/java/games/strategy/engine/lobby/server/login/LobbyLoginValidator.java#L289).
The only occurence where this object would persist is when the client is buggy and crashes half way during the authentication, or an attacked did this on purpose (requesting challenges without actually trying to solve them so the lobby shuts down)